### PR TITLE
fix: harden bash product write guard coverage

### PR DESCRIPTION
**Pull request by baksohyeon**
## Problem

On 2026-09-08, the product-path guard Bash branch still had a write-target parsing gap: `dd if=/dev/null of=<product_root>/file.txt` could pass because `of=...` was not interpreted as a destination path. The same hook also needed clearer boundaries about what it can and cannot parse so readers do not over-assume coverage.

## Why this approach

I kept the existing best-effort parser and expanded explicit write-shape detection instead of trying to evaluate shell execution semantics. A full shell interpreter model would be fragile and high risk in a PreToolUse hook, while targeted parsing plus explicit gap documentation gives deterministic enforcement for observed shapes.

## What this changes

- Update `master-ops/scripts/hooks/product-path-guard.sh` to mark Bash detection as best effort and document parser blind spots.
- Add explicit write-capable command classification for `cp`, `mv`, `rm`, `rmdir`, `mkdir`, `install`, `ln`, `touch`, `truncate`, `tee`, and `dd`.
- Parse `dd of=...` as a destination target and refuse writes that target the product path.
- Extend mutating git detection to cover mutating subcommands while allowing strict-mode read-only `git worktree list`.
- Add regression cases in `master-ops/scripts/test-product-path-guard.sh` for `dd of=...`, `git reset --hard`, and strict allowlisted `git worktree list`.
- Add `docs/internal/guard-bash-writes-report.md` with measured behavior, covered shapes, uncovered shapes, and command evidence.
- Out of scope: fully interpreting runtime shell expansion, heredoc bodies, and command substitution semantics.

## Expected effect

Direct Bash writes into product paths are refused for the expanded set of write-capable shapes, including `dd of=...` and mutating git commands such as `git reset --hard`. Reviewers can verify behavior with `bash master-ops/scripts/test-product-path-guard.sh` and see required dispatch evidence in `docs/internal/guard-bash-writes-report.md`.

## Testing

- [x] `PYTHONPATH=src python3 -m pytest tests -q` — 582 passed, 13 subtests passed in 45.23s
- [x] A test that fails without this change, or a note on why none was needed — added `bash-dd-of-equals`; before this patch that case allowed a product write
- [x] New files staged before running any scan. The scanners read tracked files, so an unstaged addition is invisible to them and the run comes back green without having looked at it.

## Review

Automated code review was run during development and no blocking findings remained in the final diff. A minor parser edge case was fixed by tracking the git subcommand position directly instead of rescanning with `index()`.

## Changelog

No root `CHANGELOG.md` entry was added because this change is a guard hardening and test/docs update in `master-ops/`.

## Template impact

This change touches `master-ops/`; existing installed operations repositories need to take this template update to get the stronger Bash write guard behavior and the updated regression coverage.

## Notes

The Bash guard remains best effort by design. Known uncovered write shapes are documented in `docs/internal/guard-bash-writes-report.md`, including runtime variable expansion where the destination path is only visible after shell evaluation.

Made with [Cursor](https://cursor.com)

---
## Summary by cubic
Hardens the Bash product-path guard so writes that the hook previously allowed are now blocked: `dd of=...` destinations, mutating git subcommands like `git reset --hard`, product writes via `--git-dir`, ambiguous `git stash` invocations, the `git apply -- --check` bypass, and `install`/`ln` source and destination edge cases. The parser stays best-effort, with covered and uncovered shapes documented in a new report.

**Bug Fixes**

- Treats `cp`, `mv`, `rm`, `rmdir`, `mkdir`, `install`, `ln`, `touch`, `truncate`, and `dd` as write-capable for product paths.
- Parses `cp`, `install`, and `ln` write targets, covering `install -d` directories, `-Sd`-style attached option values, operands after `--`, and `ln` sources.
- Keeps `git worktree list` and `git apply --check` allowed in strict mode.
- Adds regression tests for each newly blocked and allowed shape.
- Adds `docs/internal/guard-bash-writes-report.md` with measured behavior and verification evidence.
- Installed operations repositories must apply this `master-ops/` update to get the stronger guard.

<sup>Written for commit 987be1b32c5b6e17b09177da41809926899a0363. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/baksohyeon/mogui-ADE-orchestrator/pull/118?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved protection against `dd` commands writing through `of=...` targets.
  * Expanded detection of write-capable filesystem commands and mutating Git operations.
  * Improved handling of copy, installation, and symbolic-link destinations within protected product paths.
  * Continued allowing approved read-only Git commands, such as `git worktree list`, in strict mode.
  * Clarified handling of unsupported or ambiguous shell command patterns.

* **Tests**
  * Added regression coverage for blocked writes, mutating Git commands, and permitted read-only operations.

* **Documentation**
  * Documented command parsing coverage, limitations, and verification results.

### CODE REVIEW

File: docs/internal/guard-bash-writes-report.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+<exit 0>
+```
+
+```text
+$ PYTHONPATH=src python -m pytest tests -q
+582 passed, 13 subtests passed in 45.23s
+```
+
+```text
+$ master-ops/scripts/pr-body-check 1 --body-file pr-body-guard-bash-writes.md --template-file .github/pull_request_template.md
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply docs/internal/guard-bash-writes-report.md]: Updated the report to use `master-ops/scripts/pr-body-check 118 --repo baksohyeon/mogui-ADE-orchestrator`, which runs from the repository without requiring a missing body fixture file.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 2]
```diff
@@ -228,16 +233,40 @@ for parts in segments:
...
+            token=parts[sub_index]
+            if token.startswith("-"):
+                sub_index += 1
+                continue
+            git_subsub=token
+            break
     git_mutation = name == "git" and (
         (sub == "remote" and remote_action in {"rename","set-head","set-branches","update","prune","set-url","add","remove"})
         or (sub == "diff" and any(token == "--output" or token.startswith("--output=") for token in parts))
+        or (sub in git_always_mutating)
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Updated guarded-target evaluation to include `git_dir`, so mutating commands with a product-path `--git-dir` are blocked even when `--work-tree` points outside; added strict regression `strict-git-dir-product-metadata-bypass`.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 3]
```diff
@@ -228,16 +233,40 @@ for parts in segments:
...
+        or (sub == "clean" and not any(token in {"-n", "--dry-run"} for token in parts))
+        or (sub == "stash" and git_subsub not in {"list", "show"})
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Updated stash parsing so `--message` and `-m` mark subcommand parsing as ambiguous and the command is treated as mutating; added strict regression `strict-stash-message-list-bypass`.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 4]
```diff
@@ -228,16 +233,40 @@ for parts in segments:
     write_args={"-exec","-execdir","xargs","--in-place"}
     if name in {"sed","perl","ruby"}:
         write_args.add("-i")
+    git_always_mutating={
+        "add", "apply", "checkout", "cherry-pick", "commit", "merge",
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Updated `git apply` classification so `--check`, `--stat`, `--numstat`, and `--summary` remain read-only unless `--apply` is present; added strict regression `strict-allowlisted-git-apply-check`.

File: master-ops/scripts/test-product-path-guard.sh [hunk 5]
```diff
@@ -90,8 +91,11 @@ expect_allowed measured-read-only run_bash_strict "ls" "$product"
 expect_blocked strict-unmeasured-read-only run_bash_strict "cat" "$product"
 expect_blocked strict-write-capable-argument run_bash_strict "find . -exec touch file.txt \\;" "$product"
 expect_allowed strict-read-search-flag run_bash_strict "rg -i needle" "$product"
+printf 'git worktree\n' >>"$TMP/allowlist.txt"
+expect_allowed strict-allowlisted-git-worktree-list run_bash_strict "git -C $product worktree list"
 expect_blocked git-remote-mutation run_bash "git -C $product remote set-url origin https://example.invalid/repo"
 expect_blocked git-diff-output run_bash "git -C $product diff --output=result.txt"
+expect_blocked git-reset-hard run_bash "git -C $product reset --hard"
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/test-product-path-guard.sh]: Updated the reset regression to strict mode with `git reset` allowlisted, so `strict-git-reset-hard` now validates the mutating-git classifier instead of legacy read-only blocking.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 6]
```diff
@@ -228,16 +233,40 @@ for parts in segments:
...
         or (sub == "diff" and any(token == "--output" or token.startswith("--output=") for token in parts))
+        or (sub in git_always_mutating)
+        or (sub == "clean" and not any(token in {"-n", "--dry-run"} for token in parts))
+        or (sub == "stash" and git_subsub not in {"list", "show"})
+        or (sub == "worktree" and git_subsub not in {"list"})
     )
-    write_capable=any(token in write_args for token in parts[1:]) or git_mutation
+    command_write_capable = name in {
+        "cp", "mv", "rm", "rmdir", "mkdir", "install", "ln", "touch", "truncate", "tee", "dd"
+    }
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Updated write-target parsing to resolve destination operands for `cp`, `install`, and `ln`; source paths under the product root now stay readable, while product-path destinations are still blocked. Added regressions `cp-source-in-product-destination-outside`, `install-source-in-product-destination-outside`, `ln-source-in-product-destination-outside`, and `install-flagged-destination-in-product`.

File: docs/internal/guard-bash-writes-report.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+INPUT:printf ok > <caller_workspace>/file.txt
+RC:0
+STDERR:(empty)
+```
+
+### Demonstrated uncovered shape
+
+```text
+CASE:uncovered_env_expansion
+INPUT:TARGET=<product_root>/file.txt cp /dev/null $TARGET
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply docs/internal/guard-bash-writes-report.md]: Observation: the report example assigned `TARGET` only inline and expanded `$TARGET` before it was set in Bash. Remedy: rewrite the probe as two commands so expansion uses an assigned variable value. Done: the report now uses `TARGET=<product_root>/file.txt; cp /dev/null "$TARGET"` in `docs/internal/guard-bash-writes-report.md`.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 8]
```diff
@@ -216,31 +242,102 @@ for parts in segments:
...
+        elif name == "install":
+            install_value_options={"-g", "-m", "-o", "-S", "-t", "--suffix", "--target-directory", "--group", "--mode", "--owner", "--context"}
+            install_operands=collect_non_option_operands(parts, 1, install_value_options)
+            for index,token in enumerate(parts[1:],1):
+                if token in {"-t", "--target-directory"} and index + 1 < len(parts):
+                    write_targets.append(parts[index + 1])
+                elif token.startswith("--target-directory="):
+                    write_targets.append(token.split("=", 1)[1])
+            if not write_targets and len(install_operands) >= 2:
+                write_targets.append(install_operands[-1])
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: `install -d` was collecting only the final operand and missed single-operand directory creates. Remedy: detect directory mode and treat every positional operand as a destination target. Done: install parsing now extends `write_targets` with all operands for `-d` and `--directory`, with regressions `install-directory-single-target` and `install-directory-multi-target`.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 9]
```diff
@@ -216,31 +242,102 @@ for parts in segments:
...
+            break
+    apply_readonly_mode=any(token in {"--check", "--stat", "--numstat", "--summary"} for token in parts)
+    apply_force_mode=any(token == "--apply" for token in parts)
     git_mutation = name == "git" and (
         (sub == "remote" and remote_action in {"rename","set-head","set-branches","update","prune","set-url","add","remove"})
         or (sub == "diff" and any(token == "--output" or token.startswith("--output=") for token in parts))
+        or (sub in git_always_mutating)
+        or (sub == "apply" and (not apply_readonly_mode or apply_force_mode))
+        or (sub == "clean" and not any(token in {"-n", "--dry-run"} for token in parts))
+        or (sub == "stash" and (git_subsub_ambiguous or git_subsub not in {"list", "show"}))
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: `git stash show --output` was treated as read-only in strict allowlist mode even though `--output` writes a file. Remedy: mark `stash show` as mutating when `--output <path>` or `--output=<path>` is present. Done: `stash_show_writes_output` now feeds `git_mutation`, with regressions `strict-stash-show-output-flag` and `strict-stash-show-output-equals`.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 10]
```diff
@@ -216,31 +242,102 @@ for parts in segments:
...
+                write_targets.append(install_operands[-1])
+        elif name == "ln":
+            ln_operands=collect_non_option_operands(parts, 1, {"-t", "--target-directory", "-S", "--suffix"})
+            for index,token in enumerate(parts[1:],1):
+                if token in {"-t", "--target-directory"} and index + 1 < len(parts):
+                    write_targets.append(parts[index + 1])
+                elif token.startswith("--target-directory="):
+                    write_targets.append(token.split("=", 1)[1])
+            if not write_targets and len(ln_operands) >= 2:
+                write_targets.append(ln_operands[-1])
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: hard-link creation from a product-path source to an outside destination can proxy later writes into protected content. Remedy: for non-symbolic `ln`, treat source operands as guarded targets in addition to destination parsing. Done: non-symbolic `ln` now adds source operands to write-target checks, and `ln-source-in-product-destination-outside` is now a blocking regression.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 11]
```diff
@@ -216,31 +242,102 @@ for parts in segments:
...
+            if token.startswith("--message="):
+                git_subsub_ambiguous=True
+                sub_index += 1
+                continue
+            if token.startswith("-"):
+                sub_index += 1
+                continue
+            git_subsub=token
+            break
+    apply_readonly_mode=any(token in {"--check", "--stat", "--numstat", "--summary"} for token in parts)
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: `git apply -- --check` could be misread as read-only because option scanning inspected tokens after `--`. Remedy: restrict apply read-only and force-mode detection to option tokens before the first `--`. Done: apply parsing now uses pre-`--` tokens and regression `strict-apply-double-dash-check-bypass` blocks this shape in strict mode.

File: master-ops/scripts/hooks/product-path-guard.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: `install -d` semantics were incomplete and could miss product-path directory creation targets. Remedy: detect `-d` and `--directory` and include all positional operands as destination targets before guarded-path evaluation. Done: install directory mode is now covered in parser logic and by regressions `install-directory-single-target` and `install-directory-multi-target`.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 12]
```diff
@@ -216,31 +242,122 @@ for parts in segments:
...
+            for index,token in enumerate(parts[1:],1):
+                if token in {"-t", "--target-directory"} and index + 1 < len(parts):
+                    write_targets.append(parts[index + 1])
+                    ln_destination_mode=True
+                elif token.startswith("--target-directory="):
+                    write_targets.append(token.split("=", 1)[1])
+                    ln_destination_mode=True
+            if not write_targets and len(ln_operands) >= 2:
+                write_targets.append(ln_operands[-1])
+            ln_symbolic=any(token in {"-s", "--symbolic"} for token in parts[1:])
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: the previous `ln_symbolic` check read tokens past `--` and depended on a narrow flag match, which left option-boundary ambiguity around hard-link source enforcement. Remedy: remove the symbolic-link exemption and always treat `ln` source operands as guarded targets, while keeping operand parsing anchored to `--` boundaries through `collect_non_option_operands`. Done: source-side guarding now applies to all `ln` forms and `ln -s $product/source.txt $ops/link.txt` is now blocked by regression coverage.

File: master-ops/scripts/test-product-path-guard.sh [hunk 13]
```diff
@@ -100,6 +110,12 @@ expect_blocked git-diff-output-equals run_bash "git -C $ops diff --output=$produ
 expect_allowed event-log-failure-does-not-block run_bash_event_failure "ls" "$product"
 expect_blocked git-add run_bash "git -C $product add file.txt"
 expect_blocked git-work-tree run_bash "git --git-dir=$TMP/repo.git --work-tree=$product add file.txt"
+expect_allowed cp-source-in-product-destination-outside run_bash "cp $product/source.txt $ops/out.txt" "$ops"
+expect_blocked install-flagged-destination-in-product run_bash "install -b $ops/source.txt $product/out.txt" "$ops"
+expect_allowed install-source-in-product-destination-outside run_bash "install $product/source.txt $ops/out.txt" "$ops"
+expect_blocked install-directory-single-target run_bash "install -d $product/newdir" "$ops"
+expect_blocked install-directory-multi-target run_bash "install -d $ops/newdir $product/newdir2" "$ops"
+expect_blocked ln-source-in-product-destination-outside run_bash "ln $product/source.txt $ops/out.txt" "$ops"
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/test-product-path-guard.sh]: Observation: symbolic-link creation from a product-path source to an outside path leaves a write-through path into guarded content and was previously untested. Remedy: enforce source-side guarding for `ln` so both hard-link and symbolic-link source operands under the product path are blocked. Done: `ln -s $product/source.txt $ops/link.txt` is now blocked and covered by `ln-symbolic-source-in-product-destination-outside`.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 14]
```diff
@@ -216,31 +242,122 @@ for parts in segments:
...
+        elif name == "install":
+            install_value_options={"-g", "-m", "-o", "-S", "-t", "--suffix", "--target-directory", "--group", "--mode", "--owner", "--context"}
+            install_operands=collect_non_option_operands(parts, 1, install_value_options)
+            for index,token in enumerate(parts[1:],1):
+                if token in {"-t", "--target-directory"} and index + 1 < len(parts):
+                    write_targets.append(parts[index + 1])
+                elif token.startswith("--target-directory="):
+                    write_targets.append(token.split("=", 1)[1])
+            if not write_targets and install_operands:
+                if any(token in {"-d", "--directory"} for token in parts[1:]):
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: directory-mode detection for `install` scanned all tokens and could misread a literal `--directory` operand that appears after `--`. Remedy: evaluate `-d` and `--directory` only from option tokens before the first `--`. Done: directory-mode detection now uses pre-`--` option tokens, and regression `install-double-dash-literal-directory-token` confirms literal operands after `--` stay allowed.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 15]
```diff
@@ -216,31 +242,130 @@ for parts in segments:
...
+            if not write_targets and install_operands:
+                install_directory_mode=any(
+                    token == "--directory"
+                    or (
+                        token.startswith("-")
+                        and not token.startswith("--")
+                        and "d" in token[1:]
+                    )
+                    for token in option_tokens
+                )
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: `install` directory-mode detection treated any short-option token containing `d` as directory mode, so attached values like `-Sd` could be misclassified and falsely block copy-out shapes. Remedy: parse short options with value-taking semantics for `-g`, `-m`, `-o`, `-S`, and `-t`, and only treat an actual `-d` flag or `--directory` option as directory mode before `--`. Done: added `has_install_directory_mode()` and regression `install-attached-suffix-value` to keep `install -Sd ...` allowed when destination is outside the product path.

File: master-ops/scripts/hooks/product-path-guard.sh [hunk 16]
```diff
@@ -216,31 +242,130 @@ for parts in segments:
...
+                    ln_destination_mode=True
+                elif token.startswith("--target-directory="):
+                    write_targets.append(token.split("=", 1)[1])
+                    ln_destination_mode=True
+            if not write_targets and len(ln_operands) >= 2:
+                write_targets.append(ln_operands[-1])
+            if ln_destination_mode:
+                source_operands=ln_operands
+            else:
+                source_operands=ln_operands[:-1] if len(ln_operands) >= 2 else []
```
[the remark this answers is not in the document]
Comment (baksohyeon) [reply master-ops/scripts/hooks/product-path-guard.sh]: Observation: single-operand `ln` from an outside working directory leaves source operands empty in the guard and can permit hard-link write-through into a product file inode. Remedy: when destination mode is absent, treat lone `ln` operands as source operands and apply guarded-path checks. Done: updated source operand selection to use lone operands as guarded sources and added regression `ln-single-source-in-product` to block this bypass shape.
